### PR TITLE
FI-934 Add monkey patches to fhir_models to save source_content 

### LINKED ISSUE #143: When source system does not have data for Must Support data element
Thank you for reporting a possible bug in Inferno! Please fill in as much of the template below as you can.

**Subject of the issue**
When the source system does not have data for a Must Support data element such as Procedure.performed[x], a data absent reason is not passing the scenario per the general guidance.
https://www.hl7.org/fhir/us/core/general-guidance.html#must-support

**Your environment**
* Edition of inferno (Community or Program): https://inferno.healthit.gov/inferno
* Version of inferno: 1.0.2
* Which browser and version(s) is the bug present on?:  Chrome Version 83.0.4103.97

**Steps to reproduce**
Patient with a procedure that is missing performed[x] that is populated with a data-absent-reason extension

Include:

* The selected test suite: Single Patient API Procedure resource
* The URL of the FHIR server being tested https://fhir-open.sandboxcerner.com/r4/0b8a0111-e8e6-4c26-a91c-5069cbc6b1ca
* The Test ID, if it's a problem with a specific test. USCPROC-09
* Any required configuration options (client ID, client secret, etc.)
* Anything else needed to make the issue occur.

**Expected behavior**
Returning Procedure.performed[x] with data-absent-reason extension should pass
  "_performedDateTime": {
    "extension": [
      {
        "url": "http://hl7.org/fhir/StructureDefinition/data-absent-reason",
        "valueCode": "unknown"
      }
    ]
  },

Is there something wrong with how the data-absent-reason is populated?

**Actual behavior**
Tell us what happens instead. Include screenshots if possible.
Returning Procedures.performed[x] with data-absent-reason extension is failing USCPROC-09
image

### PR DESCRIPTION
**Pull request by okeefm**
Closes #143. 
This PR solves an issue where parsing resources into `fhir_models` would "drop" data that `fhir_models` doesn't support (the specific instance was with primitive extensions). 

The way this is solved is by adding in a pair of attributes on the `FHIR::Model` base class, called `source_text` and `source_hash`.

When parsing a FHIR Object out of JSON or XML, (using `from_contents`, `from_json`, or `from_xml`), the text used for that parsing will be serialized onto `source_text`.

When creating a FHIR Object out of a hash (as when creating nested FHIR Resources), the hash used to create the object will be serialized onto `source_hash`.

We also added a `source_contents` method to the `FHIR::Model` class. This method will return either: `@source_text`, if it isn't `nil`, or a JSON-ified `source_hash`, if `@source_text` is `nil`.

Pull requests into Inferno require the following items to be completed. Submitter and reviewer 
should check the relevant check boxes when the associated item is done. For items that are not 
applicable, note it's not applicable ("N/A") and check the box. For example, external Pull 
Requests do not require ticket links.

**Submitter:**
- [x] This pull request describes why these changes were made
- [x] Internal ticket for this PR: FI-934
- [x] Internal ticket links to this PR
- [x] Internal ticket is properly labeled (Community/Program)
- [x] Internal ticket has a justification for its Community/Program label
- [x] Code diff has been reviewed for extraneous/missing code
- [x] Tests are included and test edge cases
- [x] Tests/code quality metrics have been run locally and pass

**Reviewer 1:**

Name:
- [ ] Code is maintainable and reusable, reuses existing code and infrastructure
      where appropriate, and accomplishes the task's purpose
- [ ] The tests appropriately test the new code, including edge cases
- [ ] You have tried to break the code

**Reviewer 2:**

Name:
- [ ] Code is maintainable and reusable, reuses existing code and infrastructure
      where appropriate, and accomplishes the task's purpose
- [ ] The tests appropriately test the new code, including edge cases
- [ ] You have tried to break the code

### CODE REVIEW

File: test/unit/fhir_models_test.rb [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+      bundle_resource.entry.each do |entry|
+        assert entry.source_contents.present?, "entry.source_contents not populated for #{entry}"
+        assert_instance_of String, entry.source_contents
+        assert entry.resource.source_contents.present?, "entry.resource.source_contents not populated for #{entry}"
+        assert_instance_of String, entry.resource.source_contents
+      end
+    end
+
+    it 'should preserve primitive extensions in source_contents' do
+      procedure_json = File.read('test/fixtures/procedure_primitive_extension.json')
```
Comment (Jammjammjamm) [inline test/unit/fhir_models_test.rb:19]: I think we need to test preserving primitive extensions in bundle entries and contained resources.

Comment (okeefm) [reply test/unit/fhir_models_test.rb:19]: Sure, I can add that.

Comment (Jammjammjamm) [reply test/unit/fhir_models_test.rb:20]: The bundle entry test looks good, but I don't see a test for contained resources https://www.hl7.org/fhir/references.html#contained

**Comment by yunwwang** [comment]
tested with bulk data server. looks good.

File: test/unit/fhir_models_test.rb (hunk 1 above)
Comment (okeefm) [reply test/unit/fhir_models_test.rb:20]: Updated to contain a test for contained resources as well. Thanks!

**Comment by okeefm** [comment]
When this is ready to merge, I'd suggest a `squash and merge`
